src/app/zap-templates/templates/app/CHIPClientCallbacks-src.zapt

```diff
@@ -293,6 +386,201 @@ bool IMDefaultResponseCallback(const chip::app::Command * commandObj, EmberAfSta
     return true;
 }
 
+
+
+bool IMReadReportAttributesResponseCallback(const app::ReadClient * apReadClient, const app::ClusterInfo & aPath,
+                                            TLV::TLVReader * apData, uint16_t status)
+{
+    ChipLogProgress(Zcl, "ReadAttributesResponse:");
+    ChipLogProgress(Zcl, "  ClusterId: 0x%04x", aPath.mClusterId);
+
+    Callback::Cancelable * onSuccessCallback = nullptr;
+    Callback::Cancelable * onFailureCallback = nullptr;
+    NodeId sourceId                          = aPath.mNodeId;
+    // In CHIPClusters.cpp, we are using sequenceNumber as application identifier.
+    uint8_t sequenceNumber                   = static_cast<uint8_t>(apReadClient->GetAppIdentifier());
+    CHIP_ERROR err = gCallbacks.GetResponseCallback(sourceId, sequenceNumber, &onSuccessCallback, &onFailureCallback);
+
+    if (CHIP_NO_ERROR != err)
+    {
+        if (onSuccessCallback == nullptr)
+        {
+            ChipLogDetail(Zcl, "%s: Missing success callback", __FUNCTION__);
+        }
+
+        if (onFailureCallback == nullptr)
+        {
+            ChipLogDetail(Zcl, "%s: Missing failure callback", __FUNCTION__);
+        }
+
+        return true;
+    }
+
+    // struct readAttributeResponseRecord[]
+    uint16_t attributeId = aPath.mFieldId; // attribId
+    ChipLogProgress(Zcl, "  attributeId: 0x%04x", attributeId);
+    LogIMStatus(status);
+
+    if (status == EMBER_ZCL_STATUS_SUCCESS && apData != nullptr)
+    {
+        chip::TLV::TLVType attributeType = apData->GetType();
+        ChipLogProgress(Zcl, "  attributeType: 0x%02x", attributeType);
+
+        switch (attributeType)
+        {
+        case chip::TLV::kTLVType_SignedInteger: {
+            int64_t value;
+            apData->Get(value);
+            ChipLogProgress(Zcl, "  value: %" PRId64, value);
+
+            Callback::Callback<Int64sAttributeCallback> * cb =
+                Callback::Callback<Int64sAttributeCallback>::FromCancelable(onSuccessCallback);
+            cb->mCall(cb->mContext, value);
+            break;
+        }
+        case chip::TLV::kTLVType_UnsignedInteger: {
+            uint64_t value;
+            apData->Get(value);
+            ChipLogProgress(Zcl, "  value: %" PRIu64, value);
+
+            Callback::Callback<Int64uAttributeCallback> * cb =
+                Callback::Callback<Int64uAttributeCallback>::FromCancelable(onSuccessCallback);
+            cb->mCall(cb->mContext, value);
+            break;
+        }
+        case chip::TLV::kTLVType_Boolean: {
+            bool value;
+            apData->Get(value);
+            ChipLogProgress(Zcl, "  value: %s", value ? "True" : "False");
+
+            Callback::Callback<Int64uAttributeCallback> * cb =
+                Callback::Callback<Int64uAttributeCallback>::FromCancelable(onSuccessCallback);
```

Comment:
This is a part that I`m not sure. `Callback::Callback<Int64sAttributeCallback>` and `Callback::Callback<Int64uAttributeCallback>` are nice because we do not have tons of different callbacks.

On the other hand the callbacks are strongly typed at the moment and they do not need to checks that the value fits into the type that is supposed to be the expected type from the spec.

Will it be possible to keep the conversion to the right type here, and remove the changes to the callbacks types from this PR ? That will still enable IM for Read, and we can examine/discuss the type change into an other PR without loosing time ?

Reply:
I did reverted this change, and passed tests locally.
Although this is quite dangerious for calling unmatched functions, however, on the other hand, it seems to be safe for now.